### ISSUE TITLE
 reportUnusedCoroutine ignores coroutines in bool context

### ISSUE THREAD

**Issue by vehlwn** [microsoft/pyright#7853]
pyright 1.1.361
python 3.12.3

**Describe the bug**
`reportUnusedCoroutine=error` ignores missed await-s in bool contexts.

**Code or Screenshots**

```python
import asyncio

async def f() -> bool:
    return False

async def main():
    # f() # <- here pyright correctly says error: Result of async function call is not used; use "await" or assign result to variable
    if f(): # <- here it gives false negative result and don't complain at all
        print("yes")
    else:
        print("no")

if __name__ == "__main__":
    asyncio.run(main())
```

According to the python [docs](https://docs.python.org/3/library/stdtypes.html#truth-value-testing):
> An object is considered true unless its class defines either a `__bool__()` method that returns False or a `__len__()` method that returns zero.

Coroutines don't have `__bool__` or `__len__` methods so the above code enters "yes" branch which is the inverse of what would happen with `await`.

**Comment by erictraut** [comment]
The `reportUnusedCoroutine` diagnostic check is working as designed here. This isn't a false negative. This check detects cases where the return type of a coroutine call is not consumed — i.e. not passed to a function or assigned to a variable. In your case, the result of the call to `f()` is being used in a bool check.
